app/scripts/controllers/permissions/background-api.js

```diff
@@ -1,45 +1,131 @@
 import nanoid from 'nanoid';
+import { MethodNames } from '@metamask/permission-controller';
 import {
-  CaveatTypes,
-  RestrictedMethods,
-} from '../../../../shared/constants/permissions';
-import { CaveatFactories, PermissionNames } from './specifications';
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+  getEthAccounts,
+  setEthAccounts,
+  getPermittedEthChainIds,
+  setPermittedEthChainIds,
+} from '@metamask/multichain';
+import { RestrictedMethods } from '../../../../shared/constants/permissions';
+import { PermissionNames } from './specifications';
 
-export function getPermissionBackgroundApiMethods(permissionController) {
+const snapsPrefixes = ['npm:', 'local:'];
+const isSnap = (origin) =>
+  snapsPrefixes.some((prefix) => origin.startsWith(prefix));
+
+export function getPermissionBackgroundApiMethods({
+  permissionController,
+  approvalController,
+}) {
+  // To add more than one account when already connected to the dapp
   const addMoreAccounts = (origin, accounts) => {
-    const caveat = CaveatFactories.restrictReturnedAccounts(accounts);
-
-    permissionController.grantPermissionsIncremental({
-      subject: { origin },
-      approvedPermissions: {
-        [RestrictedMethods.eth_accounts]: { caveats: [caveat] },
-      },
-    });
+    let caip25Caveat;
+    try {
+      caip25Caveat = permissionController.getCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    } catch (err) {
+      // noop
```

Comment:
What types of errors are expected here? It looks like we expect an error when no permission exists for the origin, but we're not verifying that `err` represents that scenario. This could mask an unrelated error, making debugging very difficult.

I'd recommend checking the error and throwing if it's not what we expect, or reworking this to check whether the permission exists for the origin first.

Reply:
This applies to the two other instances of this below as well.

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/ac48b67c2052052fbe21e06f6a2e5fe76aaff3f7